File: backend/src/routes/master-data.ts
import { Router, Request, Response } from 'express';
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();
const router = Router();

// ─── Data Classification ─────────────────────────────────────────

router.get('/data-classification', async (_req: Request, res: Response) => {
  try {
    const data = await prisma.data_classification.findMany({
      orderBy: [{ method: 'asc' }, { sort: 'asc' }],
    });
    res.json(data.map(d => ({
      id: d.id, method: d.method, code: d.code,
      nameZh: d.name_zh, nameEn: d.name_en,
      parent: d.parent, sort: d.sort, status: d.status,
      comment: d.comment, level: d.level,
    })));
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch data classification' });
  }
});

// ─── Data Centers ────────────────────────────────────────────────

router.get('/data-centers', async (_req: Request, res: Response) => {
  try {
    const data = await prisma.data_center.findMany({ orderBy: { name: 'asc' } });
    res.json(data.map(d => ({
      id: d.id, name: d.name, createdBy: d.create_by, createdAt: d.create_at,
    })));
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch data centers' });
  }
});

// ─── Companies ───────────────────────────────────────────────────

router.get('/companies', async (req: Request, res: Response) => {
  try {
    const { search } = req.query;
    const where: any = {};
    if (search) {
      where.OR = [
        { company_code: { contains: search as string, mode: 'insensitive' } },
        { company_name: { contains: search as string, mode: 'insensitive' } },
      ];
    }

    const data = await prisma.company.findMany({ where, orderBy: { company_code: 'asc' } });
    res.json(data.map(c => ({
      id: c.id, companyCode: c.company_code, companyName: c.company_name,
      companyRemark: c.company_remark, s4: c.s4, area: c.area,
    })));
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch companies' });
  }
});

// ─── Legal Entities ──────────────────────────────────────────────

router.get('/legal-entities', async (req: Request, res: Response) => {
  try {
    const { appId } = req.query;
    const where: any = {};
    if (appId) where.app_id = appId as string;

    const data = await prisma.application_legal_entity.findMany({ where, orderBy: { create_at: 'desc' } });
    res.json(data.map(e => ({
      id: e.id, appId: e.app_id, companyCode: e.company_code, createdAt: e.create_at,
    })));
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch legal entities' });
  }
});

// ─── Help Files ──────────────────────────────────────────────────

router.get('/help-files', async (_req: Request, res: Response) => {
  try {
    const data = await prisma.help_file.findMany({ orderBy: { create_at: 'desc' } });
    res.json(data.map(f => ({
      id: f.id, usage: f.usage, fileName: f.file_name,
      filePath: f.file_path, createdBy: f.create_by, createdAt: f.create_at,
    })));
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch help files' });
  }
});

export { router as masterDataRoutes };

File: backend/app/routers/master_data.py
"""Master-data reference endpoints."""
import re
from fastapi import APIRouter, Depends, Query
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import text

from app.database import get_db

from app.auth import require_permission, require_role, Role

router = APIRouter()

UUID_RE = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", re.IGNORECASE)


# ---- Data Classification ------------------------------------------------

@router.get("/data-classification", dependencies=[Depends(require_permission("master_data", "read"))])
async def get_data_classification(db: AsyncSession = Depends(get_db)):
    try:
        result = await db.execute(
            text("SELECT * FROM eam.data_classification ORDER BY method ASC, sort ASC")
        )
        rows = result.mappings().all()
        return [
            {
                "id": r["id"],
                "method": r["method"],
                "code": r["code"],
                "nameZh": r["name_zh"],
                "nameEn": r["name_en"],
                "parent": r["parent"],
                "sort": r["sort"],
                "status": r["status"],
                "comment": r["comment"],
                "level": float(r["level"]) if r["level"] is not None else None,
            }
            for r in rows
        ]
    except Exception as e:
        print(f"Error: {e}")
        return JSONResponse(status_code=500, content={"error": "Failed to fetch data classification"})


# ---- Data Centers --------------------------------------------------------

@router.get("/data-centers", dependencies=[Depends(require_permission("master_data", "read"))])
async def get_data_centers(db: AsyncSession = Depends(get_db)):
    try:
        result = await db.execute(
            text("SELECT * FROM eam.data_center ORDER BY name ASC")
        )
        rows = result.mappings().all()
        return [
            {
                "id": str(r["id"]),
                "name": r["name"],
                "createdBy": r["create_by"],
                "createdAt": r["create_at"].isoformat() if r["create_at"] else None,
            }
            for r in rows
        ]
    except Exception as e:
        print(f"Error: {e}")
        return JSONResponse(status_code=500, content={"error": "Failed to fetch data centers"})


# ---- Companies -----------------------------------------------------------

@router.get("/companies", dependencies=[Depends(require_permission("master_data", "read"))])
async def get_companies(
    search: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
):
    try:
        params: dict = {}
        if search:
            where = "(company_code ILIKE :search OR company_name ILIKE :search)"
            params["search"] = f"%{search}%"
        else:
            where = "1=1"

        result = await db.execute(
            text(f"SELECT * FROM eam.company WHERE {where} ORDER BY company_code ASC"),
            params,
        )
        rows = result.mappings().all()
        return [
            {
                "id": r["id"],
                "companyCode": r["company_code"],
                "companyName": r["company_name"],
                "companyRemark": r["company_remark"],
                "s4": r["s4"],
                "area": r["area"],
            }
            for r in rows
        ]
    except Exception as e:
        print(f"Error: {e}")
        return JSONResponse(status_code=500, content={"error": "Failed to fetch companies"})


# ---- Legal Entities ------------------------------------------------------

@router.get("/legal-entities", dependencies=[Depends(require_permission("master_data", "read"))])
async def get_legal_entities(
    appId: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
):
    try:
        params: dict = {}
        conditions: list[str] = []

        if appId:
            if not UUID_RE.match(appId):
                return []
            conditions.append("app_id = :app_id")
            params["app_id"] = appId

        where_clause = " AND ".join(conditions) if conditions else "1=1"

        result = await db.execute(
            text(f"SELECT * FROM eam.application_legal_entity WHERE {where_clause} ORDER BY create_at DESC"),
            params,
        )
        rows = result.mappings().all()
        return [
            {
                "id": str(r["id"]),
                "appId": str(r["app_id"]),
                "companyCode": r["company_code"],
                "createdAt": r["create_at"].isoformat() if r["create_at"] else None,
            }
            for r in rows
        ]
    except Exception as e:
        print(f"Error: {e}")
        return JSONResponse(status_code=500, content={"error": "Failed to fetch legal entities"})


# ---- Help Files ----------------------------------------------------------

@router.get("/help-files", dependencies=[Depends(require_permission("master_data", "read"))])
async def get_help_files(db: AsyncSession = Depends(get_db)):
    try:
        result = await db.execute(
            text("SELECT * FROM eam.help_file ORDER BY create_at DESC")
        )
        rows = result.mappings().all()
        return [
            {
                "id": r["id"],
                "usage": r["usage"],
                "fileName": r["file_name"],
                "filePath": r["file_path"],
                "createdBy": r["create_by"],
                "createdAt": r["create_at"].isoformat() if r["create_at"] else None,
            }
            for r in rows
        ]
    except Exception as e:
        print(f"Error: {e}")
        return JSONResponse(status_code=500, content={"error": "Failed to fetch help files"})
